Sources/AuthFoundation/Utilities/CoalescedResult.swift

```diff
@@ -12,22 +12,106 @@
+                self.task = nil
+            }
+        }
+    }
+
+    /// Designated initializer.
+    /// - Parameter taskName: The name of a background task to initiate while the operation is active.
+    public init(taskName: String? = nil) {
+        self.taskName = taskName
     }
-    
-    func start(_ operation: ((T) -> Void) -> Void) {
-        operation { result in
-            self.finish(result)
+
+    /// Indicates if the asynchronous operation is being performed.
+    nonisolated public var isActive: Bool {
+        let semaphore = DispatchSemaphore(value: 0)
+        nonisolated(unsafe) var result: Bool = false
+        Task {
+            result = await self._isActive
+            semaphore.signal()
         }
+        semaphore.wait()
+        return result
     }
 
-    func finish(_ result: T) {
-        completionHandlers.forEach { completion in
-            completion(result)
+    /// Stores the previously-fetched value, if one has been fetched and its response was successful.
+    nonisolated public var value: T? {
+        let semaphore = DispatchSemaphore(value: 0)
+        nonisolated(unsafe) var result: T?
+        Task {
+            result = await self._value
+            semaphore.signal()
+        }
+        semaphore.wait()
+        return result
+    }
+
+    /// Performs an asynchronous operation, while ensuring that multiple concurrent requests do not result in multiple active tasks.
+    /// - Parameters:
+    ///   - reset: Resets any previously fetched value before starting.
+    ///   - operation: Asynchronous operation used to fetch the value.
+    ///   - willBegin: Optional closure to invoke before beginning the operation.
+    ///   - didEnd: Optional closure to invoke with the result after the operation completes.
+    /// - Returns: The value of type `T`.
+    public func perform(reset: Bool = false,
+                        operation: () async throws -> T,
+                        willBegin: () -> Void = {},
+                        didEnd: (Result<T, Error>) -> Void = { _ in }) async throws -> T
+    {
+        if reset {
+            _value = nil
+        }
+
+        if !_isActive {
+            if let value = _value {
+                return value
+            }
+
+            _isActive = true
+            task = BackgroundTask(named: taskName)
+            willBegin()
+
+            do {
+                let result = try await operation()
+                _value = result
+                complete(with: .success(result), didEnd: didEnd)
+                return result
+            } catch {
+                complete(with: .failure(error), didEnd: didEnd)
+                throw error
+            }
         }
+
+        return try await withCheckedThrowingContinuation { continuation in
+            continuations.append(continuation)
+        }
+    }
+
+    private func complete(with result: Result<T, any Error>, didEnd: (Result<T, Error>) -> Void) {
+        _isActive = false
```

Comment:
That unblocks reentrance for the errored request. Is it fine? By design?

Reply:
Yes that's correct. The CoalescedResult should only allow a single live operation at a time, coalescing multiple concurrent calls into one actual network request (or some other async operation).  But once it's complete, and all callers have been notified, a subsequent request can be made later.

Since this type is an actor, it doesn't matter that the `_isActive` is set first, since any subsequent call to `perform` will be serialized.

Typically the CoalescedResult values are stored as `let` values to limit the number of mutable values in the library, which simplifies `Sendable` conformance.